openforcefield/typing/engines/smirnoff/parameters.py

```diff
@@ -2874,7 +2903,295 @@ def create_force(self, system, topology, **kwargs):
+            off_idx = super().add_virtual_site(fn, atoms, **kwargs)
+            return off_idx
+
+        def get_openmm_virtual_site(self, atoms, mass=None):
+            originwt = np.zeros_like(atoms)
+            originwt[0] = 1.0 # first atom is origin (where the hole should be)
+
+            xdir = np.full( len(atoms), 1.0/(originwt.shape[0]-1) ) # must == 1
+            xdir[0] = -1.0 # total sum == 0
+
+            # point towards COM (use mass weights instead of 1/N weights)
+            if mass:
+                mass = np.asarray(mass)
+                xdir[1:] = mass[1:] / mass[1:].sum()
+
+            # Seems ydir and zdir don't matter for BondCharge, since
+            # from openmm, zdir = cross(xdir, ydir) and then ydir set to 
+            # cross(zdir, xdir).
+            # We therefore allow ydir == zdir == 0, and just displace along xdir
+            ydir = np.array(xdir)
+
+            # since the origin is atom 1, and xdir is a unit vector pointing
+            # towards the center of the other atoms, we want the
+            # vsite to point away from the unit vector to achieve the desired
+            # distance
+            pos  = [-self.distance, 0.0, 0.0] # pos of the vsite in local crds
+            return openmm.LocalCoordinatesSite(atoms, originwt, xdir, ydir, pos)
+
+
+    class VirtualSiteLonePairType(VirtualSiteType):
+        """A SMIRNOFF virtual site requiring plane angles
+
+        .. warning :: This API is experimental and subject to change.
+        """
+        _ELEMENT_NAME = None
+        outOfPlaneAngle = ParameterAttribute(unit=unit.degree)
+        inPlaneAngle    = ParameterAttribute(unit=unit.degree)
+
+
+        def add_virtual_site(self, fn, atoms, **kwargs):
+            args = [ atoms, self.distance, self.outOfPlaneAngle, self.inPlaneAngle ]
+            base_args = {
+                "charge_increments" : self.chargeincrement,
+                "weights"           : None,
+                "epsilon"           : self.epsilon,
+                "sigma"             : self.sigma,
+                "rmin_half"         : self.rmin_half
+            }
+            kwargs.update(base_args)
+            return fn( *args, **kwargs)
+
+
+    class VirtualSiteMonovalentLonePairType(VirtualSiteLonePairType):
+        """A SMIRNOFF monovalent lone pair virtual site type
+
+        .. warning :: This API is experimental and subject to change.
+        """
+        _ELEMENT_NAME = 'VirtualSiteMonovalentType'
+        def add_virtual_site(self, molecule, atoms):
+            fn = molecule._add_monovalent_lone_pair_virtual_site
+            return super().add_virtual_site(fn, atoms)
+
+        def get_openmm_virtual_site(self, atoms, mass=None):
+            assert len(atoms) == 3
+            originwt = np.zeros_like(atoms)
+            originwt[0] = 1.0 # 
+
+            xdir = [-1.0, 1.0, 0.0]
+            ydir = [-1.0, 0.0, 1.0]
+
+            theta = self.inPlaneAngle.value_in_unit( self.inPlaneAngle.unit)
```

Comment:
We should be explicit about which unit we convert to here -- Does numpy expect radians?

Reply:
Yes! Good catch